File: php/mostCommonWord.php
<?php // 2O(n) complexity

function mostCommonWord(string $paragraph, array $banned): string
{
    $refined = preg_replace('/[^\w\s]/', '', strtolower($paragraph));
    $words = preg_split('/\s/', $refined);

    $map = [];
    foreach($words as $word) {
        if (in_array($word, $banned)) {
            continue;
        }
        $map[$word] = array_key_exists($word, $map) ? $map[$word] += 1 : 1;
    }

    $max = 0; $mostCommonWord = '';
    foreach ($map as $k => $v) {
        if ($v > $max) {
            $max = $v;
            $mostCommonWord = $k;
        }
    }

    return $mostCommonWord;
}

echo mostCommonWord(
    'Bob hit a ball, the hit BALL flew far after it was hit.',
    ['hit']
);

File: java/src/main/java/dev/appkr/algorithm/mostCommonWord/MostCommonWord.java
package dev.appkr.algorithm.mostCommonWord;

import static java.util.Comparator.comparing;

import java.util.*;

public class MostCommonWord {

  public static void main(String[] args) {
    System.out.println(
        mostCommonWord("Bob hit a ball, the hit BALL flew far after it was hit.",
            new HashSet<String>() {{add("hit");}}));
    // NOTE in contains operation Set is O(1) while List is O(n)
    // @see https://soft.plusblog.co.kr/74
  }

  static String mostCommonWord(String paragraph, Set<String> banned) {
    String[] words = paragraph.toLowerCase().replaceAll("[^\\w\\s]", "").split("\\s");

    Map<String, Integer> map = new HashMap<>();
    for (String word : words) {
      if (banned.contains(word)) {
        continue;
      }
      if (!map.containsKey(word)) {
        map.put(word, 1);
      }
      map.put(word, map.getOrDefault(word, 1) + 1);
    }

    return map.entrySet().stream().max(comparing(Map.Entry::getValue)).get().getKey();
  }
}
